src/components/RichEditor/LanguagePlugin.js

```diff
@@ -0,0 +1,132 @@
+    if (selectedNode.nodeName === 'SPAN') {
+      if (!selectedNode.innerText.trim()) {
+        editor.dom.setOuterHTML(selectedNode, newText);
+      } else {
+        editor.dom.setOuterHTML(selectedNode, `${selectedNode.outerHTML}${newText}`);
+      }
+    } else if (!selectedNode.innerText.trim() && selectedNode.innerText !== ' ') {
+      selectedNode.innerHTML = newText;
+    } else if (selectedNode.nodeName === 'P') {
+      selectedNode.innerHTML += newText;
+    } else if (selectedNode.nodeName === 'BR') {
+      editor.dom.setOuterHTML(selectedNode, newText);
+    } else {
+      editor.dom.setOuterHTML(selectedNode, `${selectedNode.outerHTML}${newText}`);
+    }
+    const newSpan = editor.dom.get('new_span');
+    editor.selection.select(newSpan);
+    editor.selection.collapse(false);
+    newSpan.removeAttribute('id');
+  };
+  const resetLanguage = () => {
+    replaceText('<span id="new_span">&#65279</span>');
+  };
+
+  const openDialog = (buttonApi) => {
+    const selectedNode = editor.selection.getNode();
+    const state = store.getState();
+    return editor.windowManager.open({
+      title: 'Language plugin',
+      body: {
+        type: 'panel',
+        items: [
+          {
+            type: 'htmlpanel', // component type
+            html: `<div>Current language: ${languages[getLanguageforEditor(state, editor.id)].nativeName}</div>`,
+          },
+          {
+            type: 'selectbox',
+            name: 'language',
+            label: 'Language',
+            items: Object.keys(languages).map(lang => ({
+              value: lang,
+              text: languages[lang].nativeName,
+            })),
+          },
+        ],
+      },
+      buttons: [
+        {
+          type: 'cancel',
+          text: 'Close',
+        },
+        {
+          type: 'submit',
+          text: 'Save',
+          primary: true,
+        },
+      ],
+      onSubmit(api) {
+        const data = api.getData();
+        editor.focus();
+        if (!editor.selection.isCollapsed()) {
+          editor.notificationManager.open({
+            text: 'You cannot set the language of a selection. Change the language first and then continue typing.',
+            type: 'error',
+          });
+          api.close();
+          return;
+        }
+        if (data.language !== BROWSER_DEFAULT) {
+          const dummySpan = `<span lang=${data.language} id="new_span">&#65279</span>`;
+          replaceText(dummySpan);
+        } else if (selectedNode.lang) {
+          resetLanguage(selectedNode);
+        }
+        store.dispatch(updateFormLanguage(data.language, editor.id));
+        buttonApi.setActive(data.language !== BROWSER_DEFAULT);
+        api.close();
+      },
+    });
+  };
+
+  editor.ui.registry.addToggleButton('language', {
+    text: 'Foreign language',
+    onAction(buttonApi) {
+      openDialog(buttonApi);
+    },
+    onSetup(buttonApi) {
+      editor.addShortcut('Meta+L', 'Switch to default language', () => {
```

Comment:
Nit: I think it might be better just not including this... It feels like potentially weird behavior that Meta+L will take you out, but hitting it again won't put you back in. I think it is correct to not assume their language and not allowing them to Meta+L into a foreign language, but it might be better to just not include at all. Would like to hear your thoughts on this and potentially we can check in with others as well.

Reply:
I could see this either way; I figured if people are switching in and out of a language often, they would find it annoying to have to open up the menu to switch back to the default.